Relocatable OCaml - `--with-relative-libdir`

dra27 opened the pull request:
This is the second of three PRs which implement Relocatable OCaml as proposed in ocaml/RFCs#53. The series of changes in this PR combine to allow the absolute location of the Standard Library (e.g. `/usr/lib/ocaml`) to be removed from both the C runtime (`ocamlrun` and `libcamlrun.a`, etc.) and also from the `Config` module in the `ocamlcommon` compiler-libs library. The patches address sections 3 & 4 of the RFC.

The key changes are:
- A new primitive `%standard_library_default` allows an OCaml program to determine the default of the Standard Library as a compile-time constant. In particular, it means that rather than `Config.standard_library_default` being fixed _once_ when `config.ml` itself is compiled, it allows the compiler to determine its value _each_ time a program is linked.
- A new command line option, `-set-runtime-default` allows the calculated value for the `%standard_library_default` primitive to be overridden when linking an executable.
- A `configure` option, `--with-relative-libdir`, which allows the compiler to be configured to expect to find the Standard Library in a location specified relative to where the compiler is running from. When `./configure` is run with no arguments, the default location of the Standard Library is `/usr/local/lib/ocaml` and binaries are installed to `/usr/local/bin`. The equivalent relative compiler would be configured with `--with-relative-libdir=../lib/ocaml`.
- Utilisation of both the `BUILD_PATH_PREFIX_MAP` support added in #1515 and usage of `-fdebug-prefix-map`-style options to the C compiler to make the compiler's artefacts considerably more reproducible when `--with-relative-libdir` has been specified.

These changes necessitate considerable churn in the start-up routines for the bytecode runtime, and in the way `argv[0]` is being processed. While this code is in motion, there are several additional changes which aren't a strict requirement of the main change, but are here because this is all being shaken up:
- Building the mingw-w64 port using the MSYS2 improves slightly, in that `configure` now correctly recognises that it is a Cygwin-like environment, and uses `cygpath` et al as necessary.
- On Windows, if backslashes are included in `--prefix` (e.g. `./configure --prefix='C:\OCaml'`) then these are preserved in the resulting compiler (this implements a slightly more sensible version of #658). Apart from grinding my own axe where this is concerned, it prevents "mixed" slash paths (which look particularly amateur) from being generated when the compiler is configured `--with-relative-libdir`.
- There is an obscure, but nonetheless present, bug in the order in which the bytecode runtime processes the various options for starting up. In particular, if an executable has been compiled with `-custom`, it is possible to direct the resulting executable to load a different bytecode image by manipulating `argv[0]`. This issue is fixed here, and on normal systems (where `caml_executable_name` is implemented), an executable compiled with `-custom` _only_ loads the bytecode image it was compiled with and can no longer be directed to a load a different one.
- As part of improving reproducibility, the Cygwin version of `ocamlopt` now calls `as` directly, just as Linux does, rather than going via `gcc`.

<details>
<summary>Technical background</summary>
The crux of this PR is this pair of innocuous-looking lines:

https://github.com/dra27/ocaml/blob/0728f6af2aae32a97c2a7a1214c25736a26a479b/runtime/dynlink.c#L91
and
https://github.com/dra27/ocaml/blob/0728f6af2aae32a97c2a7a1214c25736a26a479b/utils/config.generated.ml.in#L23
which, by default, are:
```c
  if (stdlib == NULL) stdlib = "/usr/local/lib/ocaml";
```
and
```ocaml
let standard_library_default = {|/usr/local/lib/ocaml|}
```

Here, `/usr/local` is the _installation prefix_. A consequence of being relocatable, as defined in the RFC, is that this string cannot appear (in any encoding!) in the compiler binaries. At present, the string is present in the `Config` module and in the _bytecode_ runtime. Its presence in the bytecode runtime means that essentially _all_ bytecode executables contain the location of the OCaml Standard Library, either directly, if compiled with `-output-obj`, `-output-complete-exe`, etc., or through the `ocamlrun` executable, if compiled with default options or `-custom`. Conversely, native  executables only contain the location of the OCaml Standard Library if they link the `Config` module from the `ocamlcommon.cmxa`[^1].

For the compiler or runtime, installed to `/usr/local/bin/ocamlopt` or `/usr/local/bin/ocamlrun`, it is straightforward for `ocamlopt`/`ocamlrun` to determine the directory containing the running executable (`/usr/local/bin`) and to instead contain the location `../lib/ocaml` and thus combine the two to _derive_ `/usr/local/lib/ocaml`. In the code, `../lib/ocaml` is the _default_ value, which is _static_ and may be explicit-relative or absolute, where the calculated `/usr/local/lib/ocaml` is the _effective_ value, which is _dynamic_ and must be absolute. Thus, at module initialisation in:
https://github.com/ocaml/ocaml/blob/0728f6af2aae32a97c2a7a1214c25736a26a479b/utils/config.common.ml.in#L23-L30
we can instead calculate the _effective_ value `"/usr/local/lib/ocaml"` from the _default_  `standard_library_default = {|../lib/ocaml|}`.

If only it were _that_ straightforward! The calculation of the _effective_ location requires the caller to agree to be invoked from a binary located in a specific place relative to the effective location. The compiler is not the only consumer of `Config.standard_library`. Consider the trivial program `display.ml`:

```ocaml
Printf.printf "Config.standard_library = %S\n" Config.standard_library
```

compiled with:
```console
$ pwd
/home/dra/work
$ command -v ocamlopt
/usr/local/bin/ocamlopt
$ ocamlopt -o display -I +compiler-libs ocamlcommon.cmxa display.ml
$ ocamlopt -where
/usr/local/lib/ocaml
$ ./display
/usr/local/lib/ocaml
```

Today, the two final commands display the same result. If `Config.standard_library` always uses the _effective_ value as the default for the Standard Library location, `ocamlopt -where` will continue to display `/usr/local/lib/ocaml` but `./display` will suddenly display `/home/dra/work/../lib/ocaml` (or something similar, but nonetheless not `/usr/local/lib/ocaml`).

Finally, bytecode executables pose an additional problem. Let us extend the trivial program slightly:
```ocaml
Unix.realpath Config.standard_library
|> Printf.printf "Config.standard_library = %S\n"
```

compiled with:

```console
$ /usr/local/bin/ocamlc -o display -I +unix -I +compiler-libs unix.cma ocamlcommon.cma display.ml
```

Supposing we have _two_ installations of _the same version and configuration_ of OCaml, one in `/usr/local` and another in `~/.opam/default`. In this case, we have:

```console
$ /usr/local/bin/ocamlrun -config
standard_library_default: /usr/local/lib/ocaml
...
shared_libs_path:
  /usr/local/lib/ocaml/stublibs

$ /usr/local/bin/ocamlc -where
/usr/local/lib/ocaml

$ ~/.opam/default/bin/ocamlrun -config
standard_library_default: /home/dra/.opam/default/lib/ocaml
...
shared_libs_path:
  /home/dra/.opam/default/lib/ocaml/stublibs

$ /home/dra/.opam/default/bin/ocamlc -where
/home/dra/.opam/default/lib/ocaml
```

Now, executing `./display` involves loading `dllunixbyt.so` from the `stublibs` directory. In these two invocations:

```console
$ /usr/local/bin/ocamlrun ./display
/usr/local/lib/ocaml

$ ~/.opam/default/bin/ocamlrun ./display
/usr/local/lib/ocaml
```
there are two important things we reasonably expect to happen:
1. `./display` will display the _same_ path in each case, which will be `/usr/local/lib/ocaml` (`Config.standard_library` for the compiler it was built with)
2. Each `ocamlrun` will load `dllunixbyt.so` from its own installation - i.e. `/usr/local/bin/ocamlrun` will load `/usr/local/lib/ocaml/stublibs/dllunixbyt.so` and `~/.opam/default/bin/ocamlrun` will load `~/.opam/default/lib/ocaml/stublibs/dllunixbyt.so`

This leads to what I think is not an entirely obvious distinction. There are _two_ locations to consider: the default location of the Standard Library for the _runtime_, and the default Standard Library location for the _mutator_. `Config.standard_library_default` refers to the _mutator_ (i.e. program's view), but this may not necessarily be the same value as the _runtime_ needs for `OCAML_STDLIB_DIR` in `runtime/dynlink.c`. In practice, this only affects standalone bytecode images - i.e. the situation where the runtime executable and the bytecode image are in separate locations. In all other compilations (including native code; though the native runtime doesn't ever care about the location of the Standard Library), there are still two values, but they are always the same.
</details>

The changeset is best reviewed commit-by-commit (and, I'm afraid, armed with the "Technical background" explanation...):
- `enum caml_byte_program_mode` is augmented with a new `APPENDED` option which is used for `-custom`. `caml_main` then uses `caml_byte_program_mode` to differentiate between a `#!`-style "standalone" bytecode image running via `ocamlrun` and a `-custom` executable. While moving the code around, I renamed the `COMPLETE_EXE` enumeration constant to `EMBEDDED` as the mode is also used with `-output-obj`.
- The `runtime-launch-info` is one of two places where the _binary_ directory, rather than _library_ directory is embedded in a file. The format for `runtime-launch-info` is trivially extended to recognise `.` as referring to the directory containing the compiler (note that `runtime-launch-info` is _not_ a generally-configurable file - supporting arbitrary relative paths here would be a hypothetical installation where the runtime executables are installed to a different location from the compiler, which isn't supported or needed at present).
- A symbol `caml_runtime_standard_library_default` for the _runtime_ default value, principally so that `OCAML_STDLIB_DIR` is only referred to in once place.
- `%standard_library_default` is the most involved change, introducing a compile-time constant to retrieve the _mutator_ default value. In native code, the Standard Library location is only present if the `Config` module is linked. `ocamlopt` therefore has to create the string as part of linking an executable, which is done using a similar mechanism to the `caml_apply`, `caml_curry` and `caml_send` functions - a new field in the cmx header records that the compilation unit uses the `%standard_library_default` primitive. At link-time, if any of the compilation units has set this flag, the linker creates `caml_standard_library_nat` containing the location of the standard library and synthesises references to it. For bytecode, a similar technique is used, except that the linker _already_ has the full list of `%`-primitives which are used by the program, so there is no need for cmo format to be changed. There are various strategies on offer for exactly where and how the value is stored. It is needed to link the bytecode runtime (i.e. the value has to be included in the same places where C tables of primitives and so forth are generated). It is also has to be included in bytecode images where no C is being produced. Given that `-output-complete-exe` and so forth share the same value for both _runtime_ and _mutator_, for bytecode images I've opted to add a new `OSLD` bytecode section containing the _mutator_ value, which is read to `caml_standard_library_default` (_not_ to `caml_runtime_standard_library_default`). Then, as with the other compile-time constants, it's just a matter of adding another `caml_sys_const_` primitive in `runtime/sys.s` which returns an OCaml copy of that string. I've proposed this as a `%`-primitive, rather than a "known" C primitive for two reasons. Firstly, the behaviour of this _is_ a compile-time constant (i.e. something which the compiler must work out when linking the program), and the other compile-time constants are `%` primitives. Secondly, the standard library location only wants to be embedded when it's actually needed; if the native runtime used a C primitive, there would have to be a sentinel value or a default for `caml_standard_library_nat` - it seems worse to have a C primitive sat in the runtime which could return an invalid value (i.e. `caml_sys_const_standard_library_default` always returns a correct value in bytecode, but if exposed in native code, it would be possible to end up with a program which called it, but got a default empty string back). Given that these primitives are _never_ intended to be called by user-code (because of the difference in linking), unlike in #13465, I think it's better to have the C primitive for bytecode only, and the completely synthesise the function in ocamlopt only when needed.
- `%standard_library_default` provides a mechanism which means that the string constant carved into `utils/config.ml` when the compiler was _built_ can now be determined when the compiler is _run_. The default value determined by the compiler is simply that same absolute path. The `-set-runtime-default` provides a mechanism to change that default value when linking a specific program.
- These two features together combine to allow `Config.standard_library_default` to be changed to be the result of `%standard_library_default`. This mechanism is a necessary consequence of `%standard_library_default` for cross-compilation. When linking a cross-compiling version of `ocamlopt`, that compiler is built with a compiler which uses a host standard library but the resulting cross-compiler it's linking should default to a different _target_ standard library (see the change in `Makefile.cross`).
- At this point, `%standard_library_default` means that `"/usr/local/lib/ocaml"` has now moved the `Config` module to the executables themselves. Where before, `ocamlopt` and `ocamlc.byte` both had the string location linked in code via `Config.standard_library`, it's now instead embedded in `caml_standard_library_nat` for `ocamlopt` (and `ocamlc.opt`) and in an `OSLD` section for `ocamlc.byte`. The next commit then allows explicit-relative values of this path to be interpreted by both the runtime and compiler. This is then activated using `-set-runtime-default`, so that `caml_standard_library_nat` is changed to be `"../lib/ocaml"`. Most of the complexity here arises from the fact that the bytecode C runtime needs to be able to do this. Especially in the light of the work on `ld.conf`  in #14243, rather than implementing the logic in _both_ C and OCaml, the logic is implemented just in C and the `Config` module uses a new `caml_sys_get_stdlib_dirs` primitive, which is passed the result of `%standard_library_default` and returns both the _effective_ value and also the directory containing the executable (which is used to implement `Config.bindir` for `ocamlmklib`). Note, in passing, that `ocamlmklib` on Windows now searches for tools in the same way as on Unix, as there's no need for the PATH-search previously there. The C implementation itself is in `caml_locate_standard_library` which is implemented separately for Unix and Windows in `runtime/unix.c` and `runtime/win32.c`. The tools used to implement this (`dirname`, `realpath`, etc.) are not equivalently available on Windows, by which I mean that the functions don't have exactly equivalent semantics. In particular, an exact `dirname` is not available on Windows, but `GetFullPathName` has the required semantics for this specific use and actually has an option to return the dirname (albeit indirectly). On this occasion, it therefore seemed easier to make the entire "work out the location of the Standard Library" operation platform-specific, than to go to more effort to construct platform-specific building blocks for a generic version of this function.
- The CI matrix for pull requests is altered to test a mix of absolute/relative builds by default. An additional test is added to deal with the "cross-runtime" example the _Technical background_ above. Specifically, after building the compiler, it is duplicated to a new location, and a check is added to ensure `ocamlc.byte -where` produces the expected result regardless of whether the original or copied `ocamlrun` is used. If the `CI: Full matrix` label of #14013 is added, then each of the jobs builds an _additional_ compiler with the alternate configuration (i.e. the jobs which by default are `--without-relative-libdir` then have an additional `--with-relative-libdir` build created, and vice versa). These two compilers are likewise used for the "cross-runtime" `ocamlc.byte -where` test.
- Finally, there's a little bit of additional plumbing added to go from "Relocatable" to "Reproducible" by utilising `BUILD_PATH_PREFIX_MAP` and directly adding `-ffile-prefix-map` to our internal C flags. Ignoring `#!` lines and `RNTM` sections (which are addressed in the final PR), when building with `--with-relative-libdir`, none of the binaries in the build contain either the build path or the installation prefix on Windows (with mingw-w64) or Linux.

[^1]: Note that prior to #11996 in 5.3.0, native code executables which linked `dynlink.cmxa` also contained the location of OCaml Standard Library through the copy of the `Config` module in the `Dynlink_compilerlibs` library.

dra27 committed:
Allow libdir to be found relative to bindir

When configured with --with-relative-libdir, the runtime uses the
directory of the executable to determine the location of the Standard
Library. Thus, ocamlrun and the compilers look for ../lib/ocaml by
default.

This is implemented by changing caml_standard_library_default to be a
relative path, and then computing the actual value at startup (for
bytecode) and when queried (for native).

Executables (and objects) produced by the compiler always have an
absolute value of caml_standard_library_default. ocamlc.opt and
ocamlopt.opt are built using -set-runtime-default to force
caml_standard_library_default to be a relative value.

dra27 committed:
Add caml_runtime_standard_library_default

Previously, the bytecode runtime just used OCAML_STDLIB_DIR from
build_config.h. This value is now stored once in dynlink.o as
caml_runtime_standard_library_default.

dra27 committed:
Add -set-runtime-default

Allows the default location used by the bytecode runtime for the
Standard Library to be overridden when creating bytecode executables.

dra27 committed:
Update the installation test for enable-relative

dra27 committed:
Test --with-relative-libdir in CI

dra27 committed:
Fix the detection of Cygwin-like build environments

Both Cygwin and MSYS2 are now consistently detected on MSYS2. In
particular, this means that ./configure --prefix $PWD/install and
similar will cause the prefix to be correctly translated to a Windows
path, as already happens on Cygwin.

dra27 committed:
Preserve backslashes in --prefix

Previously, the --prefix argument was always normalised with cygpath -m
which meant that regardless of the argument, the paths used in the
compiler would always use slashes.

This behaviour is preserved if a slash is detected in the argument, i.e.
the caller explicitly uses mixed notation (e.g. `--prefix=C:/Prefix` or
`--prefix $PWD/install`). In particular, it means that a Cygwin-style
path will be correctly converted to a Windows-style path.

If the path uses backslashes, then it is still converted to use forward
slashes for the installation commands, but the backslashes are otherwise
preserved and used within the build itself.

dra27 committed:
Interpret . in runtime-launch-info

The runtime-launch-info file includes the location of the binary
directory. The compiler is extended so that . refers to the directory of
the compiler binary.

dra27 committed:
Increase reproducibility of relative artefacts

dra27 committed:
Bootstrap

dra27 committed:
Add %standard_library_default

%standard_library_default allows Config.standard_library_default to be
converted to a compile-time derived value, as with existing compile-time
constants such as %backend_type, etc. This paves the way for allowing
Config.standard_library_default to be changed at link-time, rather than
fixed when the Config module itself is compiled.

dra27 committed:
Use %standard_library_default in Config

Config.standard_library_default is now implemented using the
%standard_library_default primitive. This allows a convenient test which
can be added for `-set-runtime-default`.

The change also makes the host-like nature of of
Config.standard_library_default clearer, as the build of the
cross-compiler must now (correctly) specify the location of its (target)
Standard Library.

dra27 committed:
Use as directly on Cygwin, as on Linux

dra27 committed:
Detect but ignore -fdebug-prefix-map on mingw-w64

mingw-w64 is based on GCC, so supports -fdebug-prefix-map, but the test
for it is skipped in configure. The test is no longer skipped (which
means that Config.c_has_debug_prefix_map returns true) but the flag is
still explicitly not used by the compilers (as before).

dra27 committed:
Harden startup of -custom executables

By default, ocamlrun first tries to resolve argv[0] to determine where
the bytecode image is and then tries opening the executable image
itself. This is obviously correct for ocamlrun, when being called using
a shebang or executable header, but it's not correct for -custom
executables where we _know_ that the bytecode image should be with the
executable. To achieve this, a new mode is added to
caml_byte_program_mode (and the existing ones renamed) such that
caml_byte_program_mode is now STANDARD (for ocamlrun - the existing
behaviour), APPENDED (for -custom executables - the new behaviour) and
EMBEDDED (for -output-complete-exe/-output-obj - the original use of
it).

The mode is also set directly by the linker, rather than having a
default in libcamlrun which is then overridden by the startup code for
-output-complete-exe.

In the new APPENDED mode, if caml_executable_name is implemented (i.e.
it returns a string) then this file _must_ contain the bytecode image
and no other mechanisms are used. On platforms where
caml_executable_name is not implemented, APPENDED falls back to STANDARD
for compatibility.

Technically, this stops an argv[0] injection attack on setuid/setgid
-custom bytecode executables, although setuid should be used with
-output-complete-exe, if at all.

dra27 committed:
Add Config.as_is_cc

Indication as to whether ocamlopt assembles files via the C compiler or
by calling the assembler directly.

dra27 committed:
Bootstrap

shym commented on runtime/startup_byt.c:
```diff
@@ -480,26 +480,40 @@ CAMLexport void caml_main(char_os **argv)
 
-  proc_self_exe = caml_executable_name();
+  if (caml_byte_program_mode != APPENDED || proc_self_exe == NULL) {
+    /* First, try argv[0] (when ocamlrun is called by a bytecode program) */
+    exe_name = argv[0];
+    fd = caml_attempt_open(&exe_name, &trail, 0);
+  }
```
Naïve question: it’s unclear to me why the condition isn’t just `if (proc_self_exe == NULL)`. Shouldn’t `proc_self_exe` take always precedence over `argv[0]` when it can be used?

dra27 replied:
There are so many cases that I don't think any question can be naïve! I think some of the commit message for 5395c378c154a914945de05e95b6fca8234f15da should be in the comment here - my idea here is that when we compile a `-custom` executable, this only works if a `/proc/self/exe` mechanism is available.

The rationale behind this is because there's actually an obscure, but nonetheless present, security issue with a `-custom` executable, since you can abuse `argv[0]` to cause it to load a _different_ bytecode image from the one actually appended to it.

I haven't done it in this PR, but [gpakosz/whereami](https://github.com/gpakosz/whereami) gives us all the code needed to have `caml_executable_name` work on all platforms we support.

shym replied:
Yes, so many cases…

> my idea here is that when we compile a `-custom` executable, this only works if a `/proc/self/exe` mechanism is available.

I agree with that principle. I would have expected then that the condition would be an `&&` so that `argv[0]` is never used in `APPENDED` mode.

While I’m digging again in that code, I wonder something.
IIUC, `caml_main` is only ever called in `APPENDED` and `STANDARD` (where it opens itself to behave as `APPENDED` if it finds itself the concatenation of ocamlrun with some bytecode executable) modes, never in `EMBEDDED`. Now that you distinguish `APPENDED` from `STANDARD`, could `attempt_open` be really run only in the `APPENDED` case and the control flow be simplified?

dra27 replied:
Ah, I realise my comment above describes what I think we should be doing _after_ we know that `caml_executable_name` _always_ works on all platforms... the `||` at the moment is a relaxation for platforms where `caml_executable_name` is not implemented. So that means on Linux/macOS/native-Windows, where `proc_self_exe` will never be `NULL`, that means it can't be executed but on, say, Cygwin/*BSD, etc. there is still the fallback.

For `caml_attempt_open`, I think it's worth considering this w.r.t. #13465, as in that I'm suggesting that `EMBEDDED` _also_ go through `caml_main` (but not `caml_attempt_open`, obviously) for the `EMBEDDED` case.

shym commented on runtime/caml/startup.h:
```diff
@@ -48,14 +48,19 @@ extern int32_t caml_seek_optional_section(int fd, struct exec_trailer *trail,
+  EMBEDDED  /* bytecode embedded in C (e.g. -output-complete-exe/-output-obj) */
+};
 
 extern enum caml_byte_program_mode caml_byte_program_mode;
```
Do I understand correctly and `caml_byte_program_mode` could now be declared `const`?

dra27 replied:
Yes (good catch!)

shym commented on asmcomp/asmlink.ml:
```diff
@@ -198,6 +198,13 @@ let make_globals_map units_list ~crc_interfaces =
 let make_startup_file ~ppf_dump units_list ~crc_interfaces =
+  let need_stdlib =
+    let needs_stdlib = function
+    | ({ui_need_stdlib = true; _}, _, _) -> true
+    | _ -> false
```
```suggestion
    | ({ui_need_stdlib; _}, _, _) -> ui_need_stdlib
```
maybe?

dra27 replied:
Oops - I hope that was from it having been more complicated at some point in the past 🫣

shym commented on bytecomp/bytelink.ml:
```diff
@@ -201,7 +202,7 @@ let debug_info = ref ([] : (int * Instruct.debug_event list * string list) list)
 (* Link in a compilation unit *)
 
-let link_compunit output_fun currpos_fun inchan file_name compunit =
+let link_compunit accu output_fun currpos_fun inchan file_name compunit =
```
Nitpick: the name of the new parameter is the only one not conveying its meaning. `acc_need_stdlib` is too long, `acc_stdlib` might be enough?

dra27 replied:
Possibly, although this combines with a follow-up PR in #14247 where `fold_primitive` below is now:

```ocaml
let fold_primitive (needs_stdlib, uses_dynlink) name =
```

so it really just an accumulator being threaded?

shym commented on bytecomp/bytelink.ml:
```diff
@@ -519,6 +525,25 @@ let link_bytecode ?final_name tolink exec_name standalone =
          ~filename:final_name ~kind:"bytecode executable"
          outchan (Symtable.initial_global_table());
        Bytesections.record toc_writer DATA;
+       let standard_library_default =
+         if standalone && needs_stdlib then
+           (* -set-runtime-default *)
+           if !Clflags.standard_library_default = None then
+             Some Config.standard_library_effective
+           else
+             !Clflags.standard_library_default
+         else
+           (* -custom executables don't need OSLD sections - the correct value
+              is already included in the runtime. *)
+           None
+       in
+       begin match standard_library_default with
+       | Some value ->
+           (* OCaml Standard Library Default location *)
+           output_string outchan value;
+           Bytesections.record toc_writer OSLD
+       | None -> ()
+       end;
```
I was suprised to see this chunk that way (building an `option` to destruct it immediately) instead of something like what is in `emit_runtime_standard_library_default`, namely something like:
```suggestion
       if standalone && needs_stdlib then begin
         let standard_library_default =
           Option.value
             (* -set-runtime-default *)
             !Clflags.standard_library_default
             ~default:Config.standard_library_effective
         in
         (* OCaml Standard Library Default location *)
         output_string outchan standard_library_default;
         Bytesections.record toc_writer OSLD
       end;
       (* else: -custom executables don't need OSLD sections - the correct value is
          already included in the runtime. *)
```

dra27 replied:
Indeed - this particular part (the whole `OSLD` mechanism) went through so many iterations, I've just ended up with code that's much more obtuse than it needs to be!

shym commented on bytecomp/bytelink.ml:
```diff
@@ -590,6 +615,49 @@ let output_cds_file outfile =
+let c_string_literal_of_string s =
+  let b = Buffer.create (String.length s * 2) in
+  let utf16le = Bytes.create 4 in
+  let iter u =
```
Nitpick: rename `iter` as `escape`, maybe?

shym commented on runtime/caml/sys.h:
```diff
@@ -35,6 +35,10 @@ CAMLextern void caml_sys_init (const char_os * proc_self_exe,
 CAMLnoret CAMLextern void caml_do_exit (int);
 
+/* The default location of the Standard Library as used by the
+   %standard_library_default to find */
```
Mangled comment: just missing `ld.conf`?

shym commented on runtime/sys.c:
```diff
@@ -736,6 +736,47 @@ CAMLprim value caml_sys_const_backend_type(value unit)
+   completely in libasmrun as there are no other existing instances in the
+   native runtime where OCAML_STDLIB_DIR ends up being embedded. */
+#ifndef NATIVE_CODE
+/* If this remains unset than caml_runtime_standard_library_default is used */
```
Typo?
```suggestion
/* If this remains unset then caml_runtime_standard_library_default is used */
```

shym commented on Makefile.cross:
```diff
@@ -40,8 +40,15 @@ VPATH := + $(VPATH)
 CROSS_OVERRIDES=OCAMLRUN=ocamlrun NEW_OCAMLRUN=ocamlrun \
   BOOT_OCAMLLEX=ocamllex OCAMLYACC=ocamlyacc
+# The cross-compiler is linked as build/bin/ocamlopt -o cross/bin/ocamlopt
+# Config.standard_library_default for cross/bin/ocamlopt would be default be the
```
Typo. Maybe this was the intended comment:
```suggestion
# Config.standard_library_default for cross/bin/ocamlopt would by default be the
```

shym commented on Makefile:
```diff
@@ -885,7 +887,7 @@ flexlink.opt$(EXE): \
 	rm -f $(FLEXDLL_SOURCE_DIR)/flexlink.exe
 	$(MAKE) -C $(FLEXDLL_SOURCE_DIR) $(FLEXLINK_BUILD_ENV) \
-	  OCAMLOPT='$(FLEXLINK_OCAMLOPT) -nostdlib -I ../stdlib' flexlink.exe
+	  OCAMLOPT='$(FLEXLINK_OCAMLOPT) -nostdlib -I ../stdlib $(SET_RELATIVE_STDLIB)' flexlink.exe
```
Maybe, to shorten the line (if I’m not mistaken about `USE_STDLIB`)?
```suggestion
	  OCAMLOPT='$(FLEXLINK_OCAMLOPT) $(USE_STDLIB) $(SET_RELATIVE_STDLIB)' \
	  flexlink.exe
```

shym commented on configure.ac:
```diff
@@ -2889,10 +2961,71 @@ AS_IF([test x"$prefix" = "xNONE"],
+    AS_CASE([$build_bindir_to_libdir],
+      [./*],[libdir="$bindir${build_bindir_to_libdir[#].}"],
+      [../*],[libdir="$bindir/$build_bindir_to_libdir"],
+      [AC_MSG_ERROR([--with-relative-libdir requires a relative path])])])],
```
```suggestion
      [AC_MSG_ERROR(m4_normalize([--with-relative-libdir requires an explicit
        relative path, starting with either ./ or ../]))])])],
```

shym commented on runtime/unix.c:
```diff
@@ -542,3 +545,70 @@ void caml_plat_mem_unmap(void* mem, uintnat size)
+    /* If realpath fails, use the non-normalised path for error messages. */
+    if (resolved_candidate != NULL) {
+      caml_stat_free(candidate);
+      /* caml_realpath uses malloc */
```
```suggestion
      /* realpath uses malloc */
```

shym commented on runtime/win32.c:
```diff
@@ -1342,3 +1343,88 @@ value caml_win32_get_temp_path(void)
+    if (l > 0) {
+      /* GetFinalPathNameByHandle always returns \\?\ which needs stripping.
+         UNC paths are returned as \\?\UNC\ - in this case, reuse the \\ from
+         \\?\. */
```
Did you mean to say that the last `\` in `\\?\UNC\` is reused?

dra27 replied:
Oh yeah - you may be unsurprised that that comment relates to a different iteration where it was copying the string back on itself - I've tried to reduce all the manually-rolled string functions and switched to `caml_stat_wcsdup` instead, but missed that comment!

shym commented on utils/config.common.ml.in:
```diff
@@ -71,9 +90,20 @@ let configuration_variables () =
   p "version" version;
-  p "standard_library_default" standard_library_default;
+  p "standard_library_default" standard_library_effective;
+  p "standard_library_relative" standard_library_relative;
```
Minor remark: `Config.standard_library_relative` is a `bool` but `ocamlopt -config_var standard_library_relative` is a string; maybe it doesn’t matter though.

dra27 replied:
I haven't actually ended up using `Config.standard_library_relative` anywhere (in the compiler itself or in any of the adapted programs), but it felt worth having it as an explicitly stated property. We could indeed rename the string here. At the moment, you have a relocatable compiler if `test -n $(ocamlopt -config-var standard_library_relative)` but it could instead be relocatable if `test $(ocamlopt -config-var standard_library_raw) != $(ocamlopt -config-var standard_library_default)`, say?

shym replied:
Maybe I should have noted also that `ocamlopt -config-var standard_library_default` is `Config.standard_library_effective` rather than `...default`. I’m wary of the same names being used to stand for different things between the `Config` module and the `-config` cli option, which was really my point (having an empty value correspond to a `false` makes sense to me).

dra27 replied:
Yes, I see what you mean. Perhaps it's just better to have `Config.standard_library_default` (newly introduced in this PR) / `ocamlopt -config-var standard_library_default` (which already exists) being the "effective value" (and therefore always an absolute path) and `Config.standard_library_relative` as a `string option` - so it's `None` if the compiler is configured with an absolute libdir, and `Some "../lib/ocaml"`, etc., otherwise? `ocamlopt -config-var standard_library_relative` then keeps its current interpretation (which maps quite readily from a string option)

shym replied:
That would indeed make a lot of sense to me!

shym reviewed (commented):
I’ve spent a little bit of time reading through this impressive PR, if only because it touches upon cross compilation :-)
My reading of the test part of the PR was really perfunctory, I dug deeper in the other parts but I can’t pretend I always understand what’s at stake. The best comment I can make already is that I didn’t find anything suspicious :-)
Anyway, I thought I’d post my review as it is now, as I have a couple of small remarks that came up during that read.

shym commented on configure.ac:
```diff
@@ -619,6 +661,14 @@ AC_ARG_ENABLE([mmap-map-stack],
+AC_ARG_WITH([relative-libdir],
+  [AS_HELP_STRING([--with-relative-libdir],
+    [location of the Standard Library, specified relative to --bindir])],
+  [AS_IF([test x"$withval" = 'xno'],
+    [bindir_to_libdir=''],
```
A quality-of-life suggestion (that could wait for later):
```suggestion
  [AS_CASE([$withval],
    [no],
      [bindir_to_libdir=''],
    [yes],
      [bindir_to_libdir=../lib/ocaml],
```
(it would also need a nested `AS_CASE` to yield a Windows path when applicable in the `yes` case, I suppose).
Noticed this while building a cross-compiler to Windows with relative libdir, which seemed to work as expected (ie I cannot rename the directory in which it is installed, but IIUC only the combination with the other PRs would make that work).

dra27 replied:
This is a brilliant idea, thanks! It gets rid of the two cases in the opam file, too (I absolutely want the backslashes on Windows, but I didn't like having to do that so explicitly in the opam file)

jonahbeckford commented on file_formats/cmx_format.mli:
```diff
@@ -46,7 +46,8 @@ type unit_infos =
     mutable ui_force_link: bool;          (* Always linked *)
-    mutable ui_for_pack: string option }  (* Part of a pack *)
+    mutable ui_for_pack: string option;   (* Part of a pack *)
+    mutable ui_need_stdlib: bool}         (* caml_standard_library_nat needed *)
```
What is the release process for `.cmx` changes? In the semver world this would be a backwards-incompatibility change, and that would require a major version bump. OCaml obviously doesn't do that. Instead, I think it implicitly treats each minor version change as possibly backwards-incompatible, and explicitly bumps Misc.Magic_number when necessary??

Do we need to bump Misc.Magic_number in this commit?

nojb replied:
The magic numbers are bumped unconditionally at each release, so that compiler artifacts are _never_ compatible across different major versions of the compiler.

jonahbeckford commented on middle_end/closure/closure.ml:
```diff
@@ -1058,22 +1058,29 @@ let rec close ({ backend; fenv; cenv ; mutable_vars } as env) lam =
+      | Word_size -> cst make_const_int (8*B.size_int)
+      | Int_size -> cst make_const_int (8*B.size_int - 1)
+      | Max_wosize -> cst make_const_int ((1 lsl ((8*B.size_int) - 10)) - 1)
+      | Ostype_unix -> cst make_const_bool (Config.target_os_type = "Unix")
```
Sigh. TIL that the split between Windows and Unix are plumbed all the way down to closures. Hopefully this is just for some internal flambda optimization. (Unrelated to this PR, but just something I'll have to research later since it sounds like a barrier for Windows<->Unix cross-compilation.)

dra27 replied:
This isn't part of this PR (it's got refactored slightly to add the code to emit the link-time constant), but this shouldn't be being antagonistic to cross-compilation, as it's entirely about it! The idea here is that `Sys.os_type` (and the various other things) take on the value of the _target_ (i.e. the executable being created) rather than those of the _host_ (the compiler).

jonahbeckford commented on driver/main_args.ml:
```diff
@@ -163,6 +163,10 @@ let mk_H f =
 
+let mk_set_runtime_default f =
+  "-set-runtime-default", Arg.String f, "<param>=<value>  Set the default for \
+      runtime parameter <param> to <value>"
```
This could use a list of supported parameters. ie. `<param>: standard_library_default`

dra27 replied:
Yes, indeed - if (as I hope) we end up also doing PR 11 in #14247 (which sits independently at https://github.com/dra27/ocaml/pull/186), then this list might get a bit big for a help screen. I wonder if it would be worth pre-emptively putting the list of parameters (even when it's just one) in both the ocamlc/ocamlopt man pages and the manual and then referring to that here? (or possibly going further and having "verbose" help, but as I don't think we do that for anything else yet, that's maybe too excessive!)

jonahbeckford commented:
status: I have reviewed so far the following commits:
- Fix the detection of Cygwin-like build environments
- Preserve backslashes in --prefix
- Interpret . in runtime-launch-info
- Harden startup of -custom executables
- Add caml_runtime_standard_library_default
- Add %standard_library_default - *the change to .cmx means this PR can't be backported safely*
- Add -set-runtime-default
- Bootstrap (I didn't review the binaries)
- Use %standard_library_default in Config
- Bootstrap (I didn't review the binaries)

dra27 commented:
> * Add %standard_library_default - _the change to .cmx means this PR can't be backported safely_

What's the rationale behind the backporting comment? Even without a magic number bump (and any backports would certainly need one), there's no guarantee that the artefacts of one point release can be used with another? i.e. the magic number exists to protect marshalling and to ensure the compiler doesn't segfault, but I don't think we've ever made the stronger guarantee that, say, artefacts compiled with OCaml x.y.0 can be freely linked with artefacts compiled with x.y.1

dra27 commented:
Thank you for the reviews so far, @jonahbeckford and @shym!

MisterDA commented on asmcomp/asmpackager.ml:
```diff
@@ -217,6 +217,9 @@ let build_package_cmx members cmxfile =
     else
       Clambda (get_approx ui)
   in
+  let ui_need_stdlib =
+    List.fold_left (fun acc unit -> acc || unit.ui_need_stdlib) false units
+  in
```
```suggestion
  let ui_need_stdlib = List.exists (fun info -> info.ui_need_stdlib) units in
```

There's a similar pattern below for `ui_force_link`.

dra27 replied:
I'm not sure what possessed me to write that originally! I blame the pandemic

MisterDA commented on bytecomp/bytelink.ml:
```diff
@@ -227,46 +228,44 @@ let link_compunit output_fun currpos_fun inchan file_name compunit =
+      link_archive accu output_fun currpos_fun file_name units
+
+let link_files output_fun currpos_fun =
+  List.fold_left (link_file output_fun currpos_fun) false
```
```suggestion
  List.exists (link_file output_fun currpos_fun)
```

dra27 replied:
That's not equivalent, though - we do need to link all of the files, regardless of whether they use that symbol or not!

MisterDA replied:
Woops, I got carried away.

MisterDA commented on configure.ac:
```diff
@@ -329,12 +339,44 @@ AS_IF([test x"$target_alias" != x],
+# In MSYS2, the mingw-w64 compiler are normal host compilers, which MSYS2 makes
+# available through different Environments (in a similar to the Microsoft Visual
+# Studio Tools Command Prompts; see https://www.msys2.org/docs/environments/).
+# It is possible to use MSYS2's "Cygwin" gcc (the equivalent of compiling native
+# Cygwin), in which case $build is *-*-msys*.
```
[2025-06-20 - Replacing `x86_64-pc-msys` with `x86_64-pc-cygwin`](https://www.msys2.org/news/#2025-06-20-replacing-x86_64-pc-msys-with-x86_64-pc-cygwin)

> As part of our ongoing effort to move MSYS2 closer to Cygwin, we have now replaced the `x86_64-pc-msys` triplet with `x86_64-pc-cygwin` as the default host triplet for the MSYS environment.

Not sure if our `config.guess` / `config.sub` have caught up with that, or if they're relevant here.

dra27 replied:
I think the answer is technically "no" in both cases - however, I think it is relevant inasmuch as it would be worth noting that trying to differentiate MSYS2/Cygwin here (which it isn't, fortunately) is clearly becoming harder/unnecessary.

jonahbeckford commented:
> What's the rationale behind the backporting comment? Even without a magic number bump (and any backports would certainly need one), there's no guarantee that the artefacts of one point release can be used with another? i.e. the magic number exists to protect marshalling and to ensure the compiler doesn't segfault, but I don't think we've ever made the stronger guarantee that, say, artefacts compiled with OCaml x.y.0 can be freely linked with artefacts compiled with x.y.1

I didn't understand when the magic number was bumped. It was explained to me. Thanks; ignore the comment.

jonahbeckford commented on runtime/unix.c:
```diff
@@ -542,3 +545,70 @@ void caml_plat_mem_unmap(void* mem, uintnat size)
+  /* See Filename.generic_dirname */
+  size_t n = strlen(path) - 1;
+  char *res;
+  if (n < 0) /* path is "" */
```
This should give a compiler warning, no? `n` is an unsigned `size_t` which can never be less than zero. Think `n` should be `ssize_t` or equivalent.

jonahbeckford replied:
I reviewed the ptrdiff_t fix. LGTM.

jonahbeckford commented on runtime/win32.c:
```diff
@@ -1342,3 +1343,88 @@ value caml_win32_get_temp_path(void)
+      return caml_stat_wcsdup(stdlib_default);
+    }
+
+    CAMLassert(basename != root && Is_separator(*(basename - 1)));
```
```suggestion
    CAMLassert(basename && basename != root && Is_separator(*(basename - 1)));
```

for safety since https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfullpathnamea says:

> If lpBuffer refers to a directory and not a file, lpFilePart receives zero.

edit: Actually, what would be better is to return `stdlib_default` if `basename==0`, since `basename` is used later.

dra27 replied:
Hang on, is it worth considering the domain of the values possible here - `exe_name` _will be_ the result of `GetModuleFileName` so it cannot possibly be a directory nor - given Windows semantics - can it have been replaced by a directory in a race (because the running program can't be deleted).

So I agree it's worth the assertion checking that `basename` isn't NULL, but surely it's not worth a code path to handle that case, because it's `basename` being NULL is going to happen with a coding error (i.e. any thing outside an assertion checking this should be an unreachable code path?)

jonahbeckford replied:
yes, that is true.

jonahbeckford commented:
Looks good for me after my last two comments.

lthls commented:
@dra27 reminded me today to have a look at the middle-end part of the patch (around the `%standard_library_default` primitive). Most of it looks reasonable, but I think it should not be mixed with compile-time primitives. I would advocate a dedicated Lambda/Clambda primitive, translated in `Cmmgen` to the appropriate symbol.
If you give me a few days I should be able to send you a patch with the changes I have in mind.

dra27 commented:
Rebased - review responses to follow

dra27 commented:
I have hopefully addressed all the review comments, thank you - both the "Compare" should be working above to see the changes without the noise of the rebase and I have put each change in separate "Review" commits for squashing later.

dra27 commented:
@lthls - ah, I see what you mean (I think); so instead of compiling it away as we go into {f,c}lambda, instead introduce a new middle-end primitive (which should get rid of the hacking around with `symbol_for_global`, I think?)

lthls commented:
> @lthls - ah, I see what you mean (I think); so instead of compiling it away as we go into {f,c}lambda, instead introduce a new middle-end primitive (which should get rid of the hacking around with `symbol_for_global`, I think?)

Yes. I did give it a try last week and it turned out to be less elegant (and more work) than I expected, so I'm now inclined to let the current version stay.

damiendoligez commented on bytecomp/bytelink.ml:
```diff
@@ -590,6 +608,49 @@ let output_cds_file outfile =
+  let escape u =
+    match Uchar.to_int u with
+      (* Characters with C escape sequences *)
+    | 000 (* '\0' *) -> Buffer.add_string b "\\0"
```
This is not correct when your NUL character is followed by digits.
```suggestion
    | 000 (* '\0' *) -> Buffer.add_string b "\\000"
```

damiendoligez commented on bytecomp/bytelink.ml:
```diff
@@ -590,6 +608,49 @@ let output_cds_file outfile =
+    | 010 (* '\n' *) -> Buffer.add_string b "\\n"
+    | 013 (* '\r' *) -> Buffer.add_string b "\\r"
+    | 034 (* '\"' *) -> Buffer.add_string b "\\\""
+    | 092 (* '\\' *) -> Buffer.add_string b "\\\\"
```
You may also want to escape `?` to avoid trigraph problems 😖

dra27 replied:
I checked further with , and I think this is the case:
- The Microsoft C compiler disabled trigraphs by default in Visual Studio 2010 (OCaml needs 5.x+)
- For GCC and Clang, all our supported versions will be in either `gnu11` or `gnu17` mode, both of which expressly have trigraphs disabled

It think we can therefore avoid the support entirely (@MisterDA?)

MisterDA replied:
I agree with all of the above, and trigraphs have been removed from C23, which GCC15 defaults to.
- https://en.cppreference.com/w/c/language/operator_alternative.html#Trigraphs_.28removed_in_C23.29
- https://gcc.gnu.org/onlinedocs/cpp/Initial-processing.html

damiendoligez commented on configure.ac:
```diff
@@ -330,12 +340,45 @@ AS_IF([test x"$target_alias" != x],
+# This leads to the four patterns below, all of which imply that the build is
+# taking place on a system where Cygwin's utilities (cygpath, etc.) can be
+# expected to be found and semantics (CYGWIN=winsymlinks:native, etc.) expected
+# to apply.
```
The way I parse it, this part of the sentence is missing a verb (and maybe a few more words).

dra27 replied:
_its_ semantics (...) _be_ expected to apply.

damiendoligez commented on man/ocamlc.1:
```diff
@@ -674,6 +674,15 @@ This allows to detect match failures even if a pattern-matching was
 .TP
+.BI \-set\-runtime\-default " setting=value"
+When linking an executable, override the default value for a runtime setting.
+The currently supported settings are:
```
```suggestion
The only currently supported setting is:
```

`nroff` doesn't support nested item lists, so I think it's better to be explicit that this is a one-element list.

damiendoligez commented on man/ocamlopt.1:
```diff
@@ -593,6 +593,15 @@ Save intermediate representation after the given compilation pass. The currently
 .TP
+.BI \-set\-runtime\-default " setting=value"
+When linking an executable, override the default value for a runtime setting.
+The currently supported settings are:
```
```suggestion
The only currently supported setting is:
```
Likewise.

damiendoligez commented on runtime/caml/osdeps.h:
```diff
@@ -96,6 +96,21 @@ void *caml_plat_mem_commit(void *, uintnat);
+
+   If dirname is not NULL and stdlib_default is a relative path, a copy of the
+   directory name part of exe_name is returned. If stdlib_default is an absolute
+   path, dirname is never changed.
```
This is a bit unclear. What is `dirname` ? Why are two different things returned (one in the first paragraph, one in the second) ? Or do you mean that `dirname` is set to "a copy of the directory name part of exe_name" ?

damiendoligez commented on runtime/startup_byt.c:
```diff
@@ -480,26 +480,47 @@ CAMLexport void caml_main(char_os **argv)
+     escape hatch here to fallback to checking argv[0] if proc_self_exe is
+     NULL.
+     For STANDARD mode (i.e. the current executable is ocamlrun), argv[0] is
+     tried first, as this should be the path to shebang-script/executable
+     originally executed by the user. */
+  if (caml_byte_program_mode != APPENDED || proc_self_exe == NULL) {
```
```suggestion
     originally executed by the user. */
  CAMLassert (caml_byte_program_mode != EMBEDDED);
  if (caml_byte_program_mode != APPENDED || proc_self_exe == NULL) {
```
`EMBEDDED` is not possible here, right?

damiendoligez reviewed (commented):
I skipped the `testsuite/tools/` and `ci/` files, which are over my head. 

shym reviewed (approved):
I went through the “Review” commits and this all looks good to me: thank you for that very clean update!

dra27 commented on Changes:
```diff
@@ -66,6 +66,15 @@ Working version
   (Tim McGilchrist, review by Nick Barnes, Sadiq Jaffer and
   Gabriel Scherer)
 
+- #14244: Added --with-relative-bindir which allows the runtime and the
```
```suggestion
- #14244: Added --with-relative-libdir which allows the runtime and the
```

damiendoligez reviewed (approved):
Approved after interactive review session; a few small changes are still pending.

dra27 commented:
(rebased on to same tip as #14243)

dra27 commented:
(rebased on to 4f3478e prior to update)

dra27 commented:
That (hopefully) final rebase puts this PR on top of #14243, with various consequential updates. The review commits are now squashed in.

dra27 commented:
[precheck#1082](https://ci.inria.fr/ocaml/job/precheck/1082/) seems happy. This one looks ready to go, next - thanks, everyone!

dra27 commented:
Let's try that again - an enhanced battery of tests is running in [precheck#1084](https://ci.inria.fr/ocaml/job/precheck/1084/) as well.

dra27 commented:
Rebased to clear the merge conflict - while the double-bootstrap provides a welcome warming for my very cold office, can I cajole/goad/bribe another core dev into merging this before the next merge conflict arises? 🙂

gasche commented:
AppVeyor failed, and I have no idea why -- it fails after installing Cygwin, apparently after doing anything OCaml-related, and there is no error message in [the logs](https://ci.appveyor.com/project/avsm/ocaml/builds/53191946). Oh well, let's merge anyway.

dra27 commented:
> AppVeyor failed, and I have no idea why -- it fails after installing Cygwin, apparently after doing anything OCaml-related, and there is no error message in [the logs](https://ci.appveyor.com/project/avsm/ocaml/builds/53191946). Oh well, let's merge anyway.

It passed on trunk! 😱
